### ISSUE TITLE
pip-sync 7.0.0 uninstalls pyproject-hooks

### ISSUE THREAD

**Issue by tomazfs** [jazzband/pip-tools#1917]
pip-sync uninstalls pyproject-hooks, which is requirement of build package and indirectly requirement of pip-tools. End result is broken pip-tools installation.

#### Environment Versions

Windows 10
Python version: Python 3.11.4
pip version: pip 23.2
pip-tools version: pip-compile, version 7.0.0

#### Steps to replicate

1. Create and activate venv
2. Install pip-tools
3. Create empty requirements.txt
4. Run: pip-sync

#### Expected result

Everything up-to-date

#### Actual result

Found existing installation: pyproject_hooks 1.0.0
Uninstalling pyproject_hooks-1.0.0:
  Successfully uninstalled pyproject_hooks-1.0.0

**Comment by atugushev** [comment]
Confirmed. Thanks for the issue!

```bash
$ cat requirements.in
django

$ pip-compile
#
# This file is autogenerated by pip-compile with Python 3.11
# by the following command:
#
#    pip-compile
#
asgiref==3.7.2
    # via django
django==4.2.3
    # via -r requirements.in
sqlparse==0.4.4
    # via django

$ pip-sync -a
Would uninstall:
  pyproject_hooks
Would install:
  asgiref==3.7.2
  django==4.2.3
  sqlparse==0.4.4
Would you like to proceed with these changes? [y/N]:
```

**Comment by atugushev** [comment]
t seems this PR #1885 might have broken pip-sync. There was some refactoring on the diff (what to install and what to uninstall) calculation.

cc @siddharthab

**Comment by atugushev** [comment]
@tomazfs @AxelBjork could you test this fix #1919? Does it resolve the issue?

```
pip install git+https://github.com/atugushev/pip-tools@fix-broken-sync-diff
```

**Comment by AxelBjork** [comment]
```
$ pip install git+https://github.com/atugushev/pip-tools@fix-broken-sync-diff
Successfully built pip-tools
Installing collected packages: pip-tools
  Attempting uninstall: pip-tools
    Found existing installation: pip-tools 7.0.0
    Uninstalling pip-tools-7.0.0:
      Successfully uninstalled pip-tools-7.0.0
Successfully installed pip-tools-3.2.1.dev1039

$ pip-sync requirements.txt
Everything up-to-date
```
This works like expected again

**Comment by siddharthab** [comment]
Thank you for the fix. I should have paid more attention to canonicalizing the package name.

**Comment by atugushev** [comment]
> Thank you for the fix. I should have paid more attention to canonicalizing the package name.

@subhaminion, no worries! Unfortunately, pip-tools does not have as many integration tests for pip-sync as pip-compile does.

**Comment by tomazfs** [comment]
> @tomazfs @AxelBjork could you test this fix #1919? Does it resolve the issue?
> 
> ```
> pip install git+https://github.com/atugushev/pip-tools@fix-broken-sync-diff
> ```

This works as expected.

**Comment by atugushev** [comment]
@sshishov Thanks for pointing that out. I've added the necessary tests for each case. However, if you believe that something is missing, please feel free to open a PR.
